src/av_stubs.c

```diff
@@ -0,0 +1,717 @@
+    ret = avcodec_decode_subtitle2(dec, stream->subtitle, &stream->got_frame, packet);
+    caml_acquire_runtime_system();
+  
+    if (ret >= 0) packet->size = 0;
+  
+    av->subtitle_stream = stream;
+    frame_kind = subtitle_frame;
+  }
+  
+  if(packet->size <= 0) {
+    av_packet_unref(packet);
+  }
+  
+  if(ret >= 0) {
+    stream->got_frame = 1;
+  }
+  else if (ret == AVERROR_EOF) {
+    stream->got_frame = 0;
+    frame_kind = end_of_file;
+  }
+  else if (ret == AVERROR(EAGAIN)) {
+    frame_kind = undefined;
+  }
+  else {
+    snprintf(error_msg, ERROR_MSG_SIZE, "Error decoding %s frame (%s)", av_get_media_type_string(dec->codec_type), av_err2str(ret));
+    Raise(EXN_FAILURE, error_msg);
+  }
+    
+  return frame_kind;
+}
+
+static value read_stream_frame(av_t * av, stream_t * stream, int media_tag)
+{
+  AVPacket * packet = &av->packet;
+  int stream_index = stream->index;
+  frame_kind frame_kind = undefined;
+    
+  for(; frame_kind == undefined;) {
+    if( ! av->end_of_file) {
+
+      if(packet->size <= 0) {
+	// read frames from the file
+	caml_release_runtime_system();
+	int ret = av_read_frame(av->format, packet);
+	caml_acquire_runtime_system();
+
+	if(ret < 0) {
+	  packet->data = NULL;
+	  packet->size = 0;
+	  av->end_of_file = 1;
+	}
+	else if (packet->stream_index != stream_index) {
+	  av_packet_unref(packet);
+	  continue;
+	}
+      }
+    }
+    frame_kind = decode_packet(av, stream);
+  }
+  
+  if(frame_kind == end_of_file) {
+    return Val_int(END_OF_FILE_TAG);
+  }
+  else {
+    value ans = caml_alloc_small(1, media_tag);
+
+    Field(ans, 0) = stream->frame;
+    return ans;
+  }
+}
+
+CAMLprim value ocaml_av_read_audio(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+  
+  if( ! av->audio_stream) open_audio_stream(av);
+  
+  ans = read_stream_frame(av, av->audio_stream, AUDIO_TAG);
+
+  CAMLreturn(ans);
+}
+
+CAMLprim value ocaml_av_read_video(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+
+  if( ! av->video_stream) open_video_stream(av);
+
+  ans = read_stream_frame(av, av->video_stream, VIDEO_TAG);
+
+  CAMLreturn(ans);
+}
+
+CAMLprim value ocaml_av_read_subtitle(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+
+  if( ! av->subtitle_stream) open_subtitle_stream(av);
+
+  ans = read_stream_frame(av, av->subtitle_stream, SUBTITLE_TAG);
+
+  if(ans != Val_int(END_OF_FILE_TAG)) {
+    Field(ans, 0) = _av;
+  }
+
+  CAMLreturn(ans);
+}
+
+CAMLprim value ocaml_av_read(value _av)
+{
+  CAMLparam1(_av);
+  CAMLlocal1(ans);
+  av_t * av = Av_val(_av);
+  AVPacket * packet = &av->packet;
+  stream_t ** streams = av->streams;
+  int nb_streams = av->format->nb_streams;
+  stream_t * stream = NULL;
+  frame_kind frame_kind = undefined;
+  
+  for(; frame_kind == undefined;) {
+    if( ! av->end_of_file) {
+  
+      if(packet->size <= 0) {
+	// read frames from the file
+	caml_release_runtime_system();
```

Comment:
Indeed. And I can do this in the read_stream_frame function as well. I will have to access the stream->frame value (actually accessed in the decode_packet function) before the loop and pass the frame pointer as parameter to the decode_packet function but this will be definitely more efficient.